### PR TITLE
mvebu: add support for Turris Omnia 2019

### PR DESCRIPTION
**Pull request by kkudielka**
The latest revision of the Turris Omnia (CZ11NIC23) is being shipped with a newer U-Boot version (2019.07). This version (a) relies on the existence of `/boot.scr` and (b) stores the firmware environment at 0xF0000, instead of 0xC0000.

Let's support the new U-Boot version, by adding the necessary boot script (similar to clearfog.bootscript) to the sysupgrade image, and adding some version logic to `/etc/uci-defaults/30_uboot-envtools`.

Installation with the new U-Boot works _without_ the "medkit", since the associated rescue system already includes shells accessible via serial _and_ ssh.

At the same time, improve the installation / sysupgrade procedure for the Turris Omnia with legacy U-Boot: (a) update the environment at first boot, and not during sysupgrade; (b) write the complete required set of environment variables, not just a part of it; (c) make use of the same boot script, to have unified behaviour.

Verification status:

- Turris Omnia 2019: Two successful installations reported below.
- Turris Omnia (CZ11NIC13) with legacy U-Boot: I have successfully executed a complete OpenWrt re-installation cycle with 4-LED reflash. Users should now be able to revert to factory OS, without manually resetting the environment. Also, 7-LED rescue shell works fine.

### CODE REVIEW

**Comment by adrianschmutzler** [comment]
> (stolen from Clearfog)

Doesn't sound like we want to include it here.

**Comment by kkudielka** [comment]
> 
> 
> > (stolen from Clearfog)
> 
> Doesn't sound like we want to include it here.

What I meant to say: turris-omnia.bootscript is an almost identical of clearfrog.bootscript in the same directory, as it seems to do what we need here as well.

**Comment by tmn505** [comment]
Hi Klaus.
Does Omnia 2019 U-Boot support `setexpr`? I rarely see that enabled in vendors U-Boot.
I also wonder about `boot.scr`, I don't see this file in tarballs of rootfs provided by CZ.NIC here: https://api.turris.cz/openwrt-repo/omnia-stable.

**Comment by kkudielka** [comment]
Hi Tomasz,
To my knowledge, setexpr is being used in *all* Omnia U-Boots (at least for the 2-3-4-5-LED rescue). For instance on my CZ11NIC13:
```
bootcmd=i2c dev 1; i2c read 0x2a 0x9 1 0x00FFFFF0; setexpr.b rescue *0x00FFFFF0; if test $rescue -ge 1; then echo BOOT RESCUE; run rescueboot; else echo BOOT eMMC FS; run mmcboot; fi
```
omnia-stable is still TurrisOS 3.x.y.
According to [https://docs.turris.cz/basics/tos_versions/](https://docs.turris.cz/basics/tos_versions/) (very bottom), CZ11NIC23 is being shipped with TurrisOS 4.0.1. As far as I understand the structure of their repository, the associated medkit is here: [https://repo.turris.cz/archive/4.0.1/medkit/omnia-medkit-latest.tar.gz](https://repo.turris.cz/archive/4.0.1/medkit/omnia-medkit-latest.tar.gz).
And this is their `/boot.scr`, of course tailored to TurrisOS.
```
if part uuid ${devtype} ${devnum}:${distro_bootpart} partuuid; then
        if test "${bootfstype}" = "btrfs"; then
                rootflags="commit=5,subvol=@"
                subvol="/@"
        else
                rootflags="commit=5"
                subvol=""
        fi

        # determine if SFP is connected (need this for device tree selection)
        if gpio input gpio@71_4; then
                dtb_variant=sfp
        else
                dtb_variant=phy
        fi

        if load ${devtype} ${devnum}:${distro_bootpart} ${fdt_addr_r} ${subvol}/boot/armada-385-turris-omnia-${dtb_variant}.dtb; then
                has_dtb=1
        elif load ${devtype} ${devnum}:${distro_bootpart} ${fdt_addr_r} ${subvol}/boot/dtb; then
                has_dtb=1
        else
                setenv has_dtb 0
                echo "Cannot find device tree binary!"
        fi

        if test $has_dtb -eq 1; then
                if load ${devtype} ${devnum}:${distro_bootpart} ${kernel_addr_r} ${subvol}/boot/zImage; then
                        setenv bootargs "earlyprintk rootwait ${contract} console=ttyS0,115200 rootfstype=${bootfstype} root=PARTUUID=${partuuid} rootflags=${rootflags} rw cfg80211.freg=${regdomain}"
                        bootz ${kernel_addr_r} - ${fdt_addr_r}
                        echo "Booting zImage failed"
                else
                        echo "Cannot load kernel binary"
                fi
        fi

        env delete partuuid
fi
```

**Comment by kkudielka** [comment]
...and to be sure for the 2019 version, I checked the above-mentioned u-boot repo.
`CONFIG_CMD_SETEXPR` is enabled by default in `cmd/Kconfig`.
And it is *not* disabled in `configs/turris_omnia_defconfig` (unlike many other configs).

**Comment by n8v8R** [comment]
>CZ11NIC23 is being shipped with TurrisOS 4.0.1. As far as I understand the structure of their repository, the associated medkit is here: https://repo.turris.cz/archive/4.0.1/medkit/omnia-medkit-latest.tar.gz.

*u-boot* version that ships from factory with board revision CZ11NIC23 **is not tied to the TOS version** but to the board revision instead and as far as I know is not publicly available for download.

In the TOS repo the (older) *u-boot* image version that is available for download <sup>**[1]**</sup> works supposedly for all TO board revisions, incl. CZ11NIC23, but does not provide automatic switching between SFP and PHY. Not sure what other benefits/improvements the newer *u-boot* image version provides.
___
<sup>**[1]**</sup> https://repo.turris.cz/omnia/nor_fw/uboot-turris-omnia-uart-spl.kwb

**Comment by tmn505** [comment]
@kkudielka thanks, that clears my doubts.

**Comment by kkudielka** [comment]
Just fixed a logical error in `turris-omnia.bootscipt`. When booting the medkit, there is no OpenWrt root partition present yet (still one big BTRFS partition). Then, the first `part` command will fail, and we conditionally execute the second one.
A correct PARTUUID is probably not needed to boot the temporary ramdisk image, but later during sysupgrade.

**Comment by neheb** [comment]
I remember one issue I had with my Omnia was that whenever I would use a medkit file to flash new firmware, the upgrade procedure would fail as I had my root Omnia image on /dev/sda and not /dev/sda1. Is this still an issue?

A git grep of the openwrt tree implies no.

**Comment by kkudielka** [comment]
I'll try to summarize my understanding of the situation, independent of the U-Boot version.

- Presently, platform_do_upgrade_sdcard() attempts to install the image onto the same device, as stated by root=... in /proc/cmdline.

- A normal sysupgrade probably does what most people intend.

- 4-LED factory rescue will install the medkit on /dev/mmcblk0p1 and then boot from there with root=b301 (or root=PARTUUID=.... corresponding to the same device). So, the sysupgrade image will end up at /dev/mmcblk0. I suggest we leave it like that - predictable behaviour for normal users.

- Advanced users are free to boot the initramfs kernel with other command lines, to install the sysupgrade image wherever they want it.

**Comment by adrianschmutzler** [comment]
Just be aware that I have recently moved most of the base-files to the relevant subtargets.
Since what you do here looks to me like it's specific to Turris Omnia, it would be good if you designed your changes in a way that they stick to the relevant subtarget as well, unless they are supposed to be used by other devices.

**Comment by kkudielka** [comment]
I have noticed those changes. Indeed, the existing part of /lib/upgrade/sdcard.sh which is specific to the Turris Omnia can easily be moved somewhere else. I will make a proposal.
Just to be sure: sdcard.sh is the only base-files component I am touching, so this is it. Or do you have other things in mind?

**Comment by adrianschmutzler** [comment]
> Just to be sure: sdcard.sh is the only base-files component I am touching, so this is it. Or do you have other things in mind?

Indeed, but currently sdcard.sh remains in the non-target-specific base-files, as it's used by more than one subtarget. I just don't want to introduce additional device-specific stuff there. If this gives us the opportunity to move stuff out of it, that would be even better.

**Comment by kkudielka** [comment]
I just force-pushed an update. Board-specific sysupgrade code now is in cortexa9/base-files/lib/upgrade.

**Comment by adrianschmutzler** [comment]
> I just force-pushed an update. Board-specific sysupgrade code now is in cortexa9/base-files/lib/upgrade.

That's exactly what I thought of. Thanks!

**Comment by adrianschmutzler** [comment]
One additional thing: I'd prefer to move platform_do_upgrade_turris_omnia() into lib/upgrade/platform.sh directly, as one relatively short function IMO does not require a file of its own.

**Comment by kkudielka** [comment]
I have implemented the last request above, and re-based onto the recent mvebu commits in the master branch.
**CZ11NIC23 owners** please test, whether the [installation instructions](https://openwrt.org/toh/turris/turris_omnia#installation) work. The associated  medkit, and sysupgrade image, are available [here](https://drive.google.com/open?id=1k7RfcCWIF_A7xJzr4oRwkovRL9kP4SP_).

File: target/linux/mvebu/image/turris-omnia.bootscript [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+setexpr openwrt_rootpart ${distro_bootpart} + 1
+
+# figure out partition uuid to pass to the kernel as root=
+part uuid ${devtype} ${devnum}:${openwrt_rootpart} uuid || part uuid ${devtype} ${devnum}:${distro_bootpart} uuid
+
+# generate bootargs (rootfs)
+setenv bootargs earlyprintk root=PARTUUID=${uuid} rootfstype=auto rootwait
+
+# add console= option to bootargs, if any
+if test -n "${console}"; then
```
Comment (neheb) [inline target/linux/mvebu/image/turris-omnia.bootscript]: I assume

[ -n "${console}" ] && setenv bootargs ${bootargs} console=${console}

also works.

**Comment by neheb** [comment]
Is this uboot available anywhere? I found this page: https://api.turris.cz/openwrt-repo/omnia-test/ but it seems to be the old one:

```
strings uboot-turris-omnia-uart-spl.kwb | grep 'U-Boot 2015.10-rc2'
U-Boot 2015.10-rc2 (Feb 02 2020 - 09:42:08 +0100)
```

**Comment by kkudielka** [comment]
> Is this uboot available anywhere?

U-Boot &  rescue images are coupled, and distributed in [TurrisOS packages](https://repo.turris.cz/hbl/omnia/packages/turrispackages/) these days. Look for `u-boot-omnia` and `rescue-image`. I think in `omnia-support` there's even a script, to flash the images.

**Attention:** [This CZ.NIC post](https://forum.turris.cz/t/how-to-flash-a-new-u-boot-on-turris-omnia/11826/18) suggests that the U-Boot images currently deployed are **not** compatible with CZ11NIC13.

File: target/linux/mvebu/image/turris-omnia.bootscript (hunk 1 above)
Comment (kkudielka) [reply target/linux/mvebu/image/turris-omnia.bootscript]: I don't see an advantage in making this different from clearfog.bootscript in the same directory.

**Comment by n8v8R** [comment]
Supposedly the *u-boot* version for the CZ11NIC13 board only ships from factory but is currently not available for download from the TOS repo since it may cause issues with other board revisions.

Current *u-boot* TOS builds, other than that particular for CZ11NIC13, are located at https://repo.turris.cz/omnia/nor_fw/

The *u-boot* version available for download is however compatible with CZ11NIC13 but does not provide the automatic switching between metal and SFP at boot time.

**Comment by neheb** [comment]
Ah bummer. Would be interesting to see if it can be updated.

**Comment by n8v8R** [comment]
The TO developers' intention is to provide eventually an updated *u-boot* version that unifies all board revisions but when that will happen is unknown.

Short of a user with this board extracting the *u-boot* version that ships from factory maybe the TO developers could provide it for testing purposes.

**Comment by kkudielka** [comment]
I have managed to harmonize the boot procedure between all Turris Omnia versions. `/boot.scr` is now sourced by all of them. At the same time, the temporary installation of a `medkit` is pointless and deprecated on the Turris Omnia 2019: It already ships with a rescue mode that offers SSH access. See the commit message for Turris Omnia 2019 (CZ11NIC23) installation instructions.

Test status:
- I verified that my CZ11NIC13 properly boots with the new environment and the new boot script.
- I cannot verify whether the CZ11NIC23 SSH rescue mode really allows to flash the sysupgrade image with `zcat` and `dd`.

**Comment by neheb** [comment]
My old Omnia boots fine as well.

**Comment by testeaxeax** [comment]
@kkudielka I built your patched version and successfully flashed my Turris Omnia 2019 with it.
First I tried using the medkit option but the system didn't boot. Then I tried the 5-LED SSH login and just used dd to copy the decompressed sysupgrade image into mmcblk0 and that booted successfully.

**Comment by kkudielka** [comment]
@testeaxeax Thanks for the confirmation. That's exactly the expected behaviour, as stated in the commit message.
I promise to update the wiki, should this get accepted.

**Comment by kkudielka** [comment]
I just rebased the branch, including one cosmetic change to the mmcboot environment variable.
From my point of view, this is tested on both CZ11NIC13 and CZ11NIC23, and ready for merge.

**Comment by kkudielka** [comment]
Apart from rebasing, I made one more change to the Turris Omnia (< 2019) U-Boot environment:

- Incorporate _all_ variables of the default environment defined in the factory U-Boot image. This makes the installation process more robust against corner cases (such as environment store with invalid CRC).

**Comment by kkudielka** [comment]
@adrianschmutzler I have updated the summary at the top with the current status, including verification. Any more comments? Do you think this can be merged?

**Review by ghost — COMMENTED** [review]
`mmcboot` seems to be conflicting with `mmc_boot`

`mmcboot btrload` -> `btrload` is not a valid command and produces

>Unknown command 'btrload' - try 'help'

**Comment by kkudielka** [comment]
> Unknown command 'btrload' - try 'help'

This should not happen. The old U-Boot 2015.10-rc2 provided by the vendor _does_ have btrload, and _only_ in that case 35_turris-omnia_uboot-env should update the U-Boot environment and set mmcboot. For all other U-Boot versions, we leave the environment untouched.

So, something must have gone wrong. May I ask:

- What U-Boot version do you have? E.g. find out with `strings /dev/mtd0 | grep 'U-Boot 20'`.
- What is your firmware environment now? E.g. `fw_printenv`.
- How exactly did you flash the image, and what was on the eMMC before the flash?

**Comment by n8v8R** [comment]
Tested on board revision CZ11NIC23 with vendors *u-boot* development version 

>U-Boot SPL 2019.07 (May 19 2020 - 20:45:05 +0000)

that has [default env.txt](https://github.com/openwrt/openwrt/files/4735820/default.env.txt)

for that purpose took inspiration from your code and added the `mmcboot` portion but has now been removed again. Pardon the red herring. :face_with_head_bandage: 
____
One other thing though, and that might be different between board | _u-boot_ revisions is `i2c dev 1` which on this board is incorrect, instead it is `i2c dev 2`

**Comment by kkudielka** [comment]
> One other thing though, and that might be different between board | _u-boot_ revisions is `i2c dev 1` which on this board is incorrect, instead it is `i2c dev 2`

Good spotting. I'll have a look at this.

**Comment by n8v8R** [comment]
Thanks for checking/feedback.

Hope that is not bothersome, just wondering about the duplicate `uuid` here:

`part uuid ${devtype} ${devnum}:${rootpart} uuid`

e.g. `part uuid mmc 0` prints the uuid in the console but `part uuid mmc 0 uuid` does not.

**Comment by n8v8R** [comment]
Probably outside this PR, just in case of interest - automatic switch through *turris-omnia.bootscript * between SFP and copper, provided that for each a **.dtb* is given:

```
# determine if SFP is connected (need this for device tree selection)
        if gpio input gpio@71_4; then
                dtb_variant=sfp
        else
                dtb_variant=phy
        fi

        if load ${devtype} ${devnum}:${distro_bootpart} ${fdt_addr_r} ${subvol}/boot/<filename>${dtb_variant}.dtb; then
                has_dtb=1
        elif load ${devtype} ${devnum}:${distro_bootpart} ${fdt_addr_r} ${subvol}/boot/dtb; then
                has_dtb=1
        else
                setenv has_dtb 0
                echo "Cannot find device tree binary!"
        fi

        if test $has_dtb -eq 1; then
```

**Comment by kkudielka** [comment]
> Probably outside this PR, just in case of interest - automatic switch through _turris-omnia.bootscript_ between SFP and copper, provided that for each a *_.dtb_ is given:

Something similar already is in my [playground](https://github.com/kkudielka/openwrt/commits/omnia-playground) branch (have a look at the SFP commit). It would be the next thing on my TODO list, should this PR get accepted.

**Comment by B08Z** [comment]
> 
> 
> I have managed to harmonize the boot procedure between all Turris Omnia versions. `/boot.scr` is now sourced by all of them. At the same time, I removed changes to the `medkit`, which would get complicated to catch all cases: Turris Omnia 2019 already ships with a rescue mode that offers SSH access.
> 
> **See the commit message for Turris Omnia 2019 (CZ11NIC23) installation instructions.**
> 
> Test status:
> 
>     * I verified that my CZ11NIC13 properly boots with the new environment and the new boot script.
> 
>     * I cannot verify whether the CZ11NIC23 SSH rescue mode really allows to flash the sysupgrade image with `zcat` and `dd`.

Hi KKudielka

Sorry I am a total NOOB but have some experience with linux. 
I have a new Turris Omnia 2019 (CZ11NIC23) confrimed to have the new U-Boot and have followed your instructions to flash via 4 LED Reset -> SSH root@192.168.1.1 ... However the zcat command is not recognised and so `zcat /mnt/openwrt-...-sysupgrade.img.gz | dd of=/dev/mmcblk0 bs=4096 conv=fsync` fails. 

`strings` is also no prestent but `cat /dev/mtd0 | grep "U-Boot 20"` will give you a similar (although messy) output and confimred that I have 2019.07 U-Boot. 

I am really keen to get this router on Openwrt, so anything you can suggest would be great.

Thanks.

**Comment by kkudielka** [comment]
Hi,

- please use 5-LED or 7-LED reset, to boot the rescue image. 4-LED reset does not what you want, see [Omnia Playground instructions](https://github.com/kkudielka/openwrt/releases/latest).
- after login, please make absolutely sure that you are in the rescue system, and not in TurrisOS. Maybe the login message gives a hint.
- As `zcat`replacement, you may try `gzip -dc`. Or, you may `gunzip` the compressed sysupgrade image, before copying it on your USB flash drive. And then just use `dd if=/mnt/openwrt-...-sysupgrade.img of=/dev/mmcblk0 bs=4096 conv=fsync` on the target device.
Please let me know which variant works, so that I can correct the procedure.

**Comment by B08Z** [comment]
Hey @kkudielka 

TL;DR Problem fixed - there was no problem

- Thanks very much for the quick reply.
- I conducted a 5 LED rest and then followed the instructions on Omnia Playgroud but instead of zcat simpley used `dd if=/mnt/openwrt......sysupgrade.img of=/dev/mmcblk10....` 
- A very minor note is that I had to `mkdir /mnt` before using `mount /dev/sda1`.

It would apear that the first time I did it (noting that it was a 5 LED reset just a typo in my last comment) It was successfull but I failed to take note that snapshot images do not have luci and so my quick check to see if there was a Luci front end at 192.168.1.1 were obviously not going to work.

Confirmed to work, all up and running, great effort thanks very much. I hope to see it merged in the Openwrt master release soon. 
If I can test anything for you please let me know.

**Comment by kkudielka** [comment]
Updated flash instructions according to feedback.

**Comment by vazhnov** [comment]
I'm going to buy Turris Omnia 2020 / CZ11NIC23.
@kkudielka , thank you for your work!
Any progress here, to have some support in mainline OpenWRT?

**Comment by kkudielka** [comment]
> Any progress here, to have some support in mainline OpenWRT?

At the moment nobody seems interested in merging this.

**Comment by neheb** [comment]
The core team prefers patches on the mailing list.

**Comment by adschm** [comment]
> The core team prefers patches on the mailing list.

That is certainly true for some areas, but I would not expect it to help much on this particular PR, as it only becomes harder to review there. To me, it just seems that nobody appears to be specifically interested in mvebu at the moment.

**Comment by adschm** [comment]
Since it's a no-brainer, I've picked the uboot-envtools patch and merged it, so only two remain for you to get in.

File: target/linux/mvebu/cortexa9/base-files/etc/uci-defaults/35_turris-omnia_uboot-env [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	fw_setenv -s - <<-"EOF"
+		baudrate 115200
+		bootdelay 3
+		ethact neta2
+		fdt_high 0x10000000
+		initrd_high 0x10000000
+		bootargs earlyprintk console=ttyS0,115200 rootfstype=btrfs rootdelay=2 root=b301 rootflags=subvol=@,commit=5 rw
+		bootcmd i2c dev 1; i2c read 0x2a 0x9 1 0x00FFFFF0; setexpr.b rescue *0x00FFFFF0; if test $rescue -ge 1; then echo BOOT RESCUE; run rescueboot; else echo BOOT eMMC FS; run mmcboot; setenv bootargs; run distro_bootcmd; fi
+		rescueboot i2c mw 0x2a.1 0x3 0x1c 1; i2c mw 0x2a.1 0x4 0x1c 1; mw.l 0x01000000 0x00ff000c; i2c write 0x01000000 0x2a.1 0x5 4 -s; setenv bootargs "$bootargs omniarescue=$rescue"; sf probe; sf read 0x1000000 0x100000 0x700000; bootz 0x1000000
+		mmcboot btrload mmc 0 ${kernel_addr_r} boot/zImage @ && btrload mmc 0 ${fdt_addr_r} boot/dtb @ && setenv bootargs ${bootargs} cfg80211.freg=${regdomain} && bootz ${kernel_addr_r} - ${fdt_addr_r}
```
Comment (tmn505) [inline target/linux/mvebu/cortexa9/base-files/etc/uci-defaults/35_turris-omnia_uboot-env:23]: Wouldn't it be better if You replaced `btrload` with `load`? That way users could format eMMC with whatever file system their U-Boot supports. Could be that the old U-Boot might not support this aliasing, that's why the question.

Comment (kkudielka) [reply target/linux/mvebu/cortexa9/base-files/etc/uci-defaults/35_turris-omnia_uboot-env:23]: Quote from the commit message:

> The first 9 lines are a copy of the default environment of the old U-Boot release - only modified, to run "distro_bootcmd", in case "mmcboot" fails to boot the factory OS.

So, `mmcboot` (9th of the first 9 lines) is only there to boot the factory OS (TurrisOS), which is and has always been on BTRFS. In all other cases, `mmcboot` will not be successful and fall through. Then, `distro_bootcmd` takes over, and searches for a boot script instead.

Honestly, I don't see the point in moving from `btrload` to `load` here. Especially, since the normal `load` won't allow to specify the subvolume...

Comment (tmn505) [reply target/linux/mvebu/cortexa9/base-files/etc/uci-defaults/35_turris-omnia_uboot-env:23]: > Quote from the commit message:
> 
> > The first 9 lines are a copy of the default environment of the old U-Boot release - only modified, to run "distro_bootcmd", in case "mmcboot" fails to boot the factory OS.
> 
> So, `mmcboot` (9th of the first 9 lines) is only there to boot the factory OS (TurrisOS), which is and has always been on BTRFS. In all other cases, `mmcboot` will not be successful and fall through. Then, `distro_bootcmd` takes over, and searches for a boot script instead.

Indeed, there's no point if mmcboot fails and something better takes over (I hate web interfaces where I have to scroll vertically).

> 
> Honestly, I don't see the point in moving from `btrload` to `load` here. Especially, since the normal `load` won't allow to specify the subvolume...

ACK.

**Comment by flyn-org** [comment]
I just bought a Turris Omnia that has U-Boot 2019.07.

I have tried to boot the stock OpenWrt builds of master using the ["4 LEDs"](https://docs.turris.cz/hw/omnia/rescue_modes/) update mechanism and a USB drive. This fails, seemingly leaving me with a router that does not boot.

I had more success with @kkudielka's builds, which are documented at https://github.com/kkudielka/openwrt/releases/tag/2020-09-13. I used `gunzip` to decompress the sysupgrade image therein, I `scp`'ed the result to the Turris after booting with ["5 LEDs"](https://docs.turris.cz/hw/omnia/rescue_modes/), and I used `dd` to write the image to `/dev/mmcblk0`. This booted into OpenWrt successfully.

Just to check, I applied the same `gunzip`/`scp`/`dd` process using an image I built from OpenWrt upstream master. This failed as with my first attempt.

So, it appears @kkudielka's work is necessary and sufficient to get OpenWrt to boot on the newest Turris Omnias. Like a few other folks who commented here, I am interested in seeing this merged. I am happy to help test.

**Comment by matco** [comment]
I just installed OpenWRT successfully thanks to this work of @kkudielka. I am interested in seeing this merged.

**Comment by flyn-org** [comment]
See also http://lists.openwrt.org/pipermail/openwrt-devel/2020-October/031872.html.

**Comment by tmn505** [comment]
@adschm Given the amount of feedback wouldn't it be fine merging this PR?

**Comment by adschm** [comment]
> @adschm Given the amount of feedback wouldn't it be fine merging this PR?

Given that there is 0 Reviewed-by, 0 Tested-by, and that I have no idea about boot with mvebu, I don't think so.

**Comment by tmn505** [comment]
So comments on GitHub do not count as valid without dropping email address?

**Comment by adschm** [comment]
> So comments on GitHub do not count as valid without dropping email address?

Doing text search on the page for "review" and "tested", I find 0 comments equivalent to a reviewed-by and 1 for a tested-by. The latter however includes additional information that makes it hard for me to judge whether it actually can be considered a Tested-by after all. I certainly won't read through all the comments and then try to deduce whether this is fine or not from _comments_ on a subject I don't have experience with.

So, instead of discussing about how the comments could be counted, it would be better and quicker to persuade the people having made them to provide Reviewed-by/Tested-by that actually can be "counted". Or you may look for a committer actually familiar with the subject ...

**Comment by tmn505** [comment]
@adschm You can add my Reviewed-by.

**Comment by flyn-org** [comment]
@kkudielka and @adschm, I am happy to add my Tested-by. I referenced my testing in two comments above. My test involved installing this on a "2020" Omnia, and using the result as a production router. Look for "W. Michael Petullo" in OpenWrt's commit log for my identifying information.

**Review by PsychoGame — APPROVED** [review]
Tested on Omnia 2019 and 2020, which I use as daily driver in my home network. Installs perfectly on my devices with this patchwork, and so far have not encountered any problems in my use cases.

**Review by PsychoGame — APPROVED** [review]
Sorry, by accident commented twice on the same commit.

**Comment by kkudielka** [comment]
FYI, in the last force-push I added Reviewed-by @tmn505 and Tested-by @flyn-org, as per the comments above. Hope this helps.

**Comment by aparcar** [comment]
Merged, thanks!

**Comment by flyn-org** [comment]
Thank you, @kkudielka!
